.drone.yml

```diff
@@ -32,20 +41,37 @@ steps:
             docker pull "$DOCKER_IMAGE:develop"
             docker tag "$DOCKER_IMAGE:develop" "$DOCKER_IMAGE:$DOCKER_TAG"
           else
-            cd "$${PROJECT_DIRECTORY[$project]}"
             if [[ "$project" == "km" ]]; then
-              docker build -f Dockerfile.km --build-arg GIT_BRANCH_CORE=${DRONE_BRANCH} --build-arg SGX_MODE=HW -t "$DOCKER_IMAGE:$DOCKER_TAG" --no-cache .
-            elif [[ "$project" == "core" ]]; then
-              docker build --build-arg GIT_BRANCH_CORE=${DRONE_BRANCH} --build-arg SGX_MODE=HW -t "$DOCKER_IMAGE:$DOCKER_TAG" --no-cache .
+              make clone-km BRANCH=${DRONE_BRANCH}
+              make build-km DOCKER_TAG=$DOCKER_TAG
             else
-              docker build --build-arg "$${GIT_BRANCH_ARG[$project]}=${DRONE_BRANCH}" -t "$DOCKER_IMAGE:$DOCKER_TAG" --no-cache .
+              if [[ "$MATCHING_BRANCH_P2P" -eq 1 ]]; then
+                make clone-p2p BRANCH=${DRONE_BRANCH}
+              else
+                make clone-p2p BRANCH=develop
+              fi
+              if [[ "$MATCHING_BRANCH_CORE" -eq 1 ]]; then
+                make clone-core BRANCH=${DRONE_BRANCH}
+              else
+                make clone-core BRANCH=develop
+              fi
+              make build-worker DOCKER_TAG=$DOCKER_TAG
             fi
-            cd ..
           fi
         done'
-      - cd enigma-contract && docker build --build-arg GIT_BRANCH_CONTRACT=${DRONE_BRANCH} -t enigmampc/enigma_contract:$DOCKER_TAG --no-cache . && cd ..
+      - make clone-contract BRANCH=${DRONE_BRANCH}
+      - |
+        if [[ "$MATCHING_BRANCH_INTEGRATION" -eq 1 ]]; then
+          # which includes master and develop because these always exist
+          make clone-client BRANCH=${DRONE_BRANCH}
+        else
+          # otherwise we are on a feature branch, and we'll build against develop
+          make clone-client BRANCH=develop
+        fi
+      - make build-contract DOCKER_TAG=$DOCKER_TAG
+      - make build-client DOCKER_TAG=$DOCKER_TAG
       - export NODES=3
-      - docker-compose -f docker-compose.yml -f docker-compose.hw.yml -f docker-compose.test.yml up --scale core=$NODES --scale p2p=$NODES --exit-code-from client && export RESULT=$? || export RESULT=$?
+      - SGX_MODE=HW docker-compose -f docker-compose.yml -f docker-compose.hw.yml up --scale worker=$NODES --exit-code-from client & sleep 60 && docker-compose exec -T client make test && export RESULT=$? || export RESULT=$?
```

Comment:
1. Why SGX_MODE=HW?
2. Why are all these commands on one line?(i realise it was like this before)
3. you can replace the `&& export RESULT=$? || export RESULT=$?` with just `; export RESULT=$?`

```
- docker-compose -f docker-compose.yml -f docker-compose.hw.yml up --scale worker=$NODES --exit-code-from client &
- sleep 60
- docker-compose exec -T client make test; export RESULT=$?
```

Reply:
1. Because this is our DroneCI where we run the tests in Hardware Mode, so we need to set `SGX_MODE=HW` *AND* run the `_hw` docker images.

2. Not sure, because it works for now... (but not a strong argument)

3. Nope ;) If I do as you say, a failed command will abort the run. The way I am doing it, the long command will always return `true` and capture the failed result in that variable. I need to do this so that I can clean up all the resources, and return the success|failed result after clean up

Reply:
1. alright :+1: 
2. sure
3. huh. I've never written `.drone.yml` files before so good to know.